### ISSUE TITLE
inla.mesh.2d error "Error in qr.coef(a, b)"

### ISSUE THREAD

**Issue by gaiteiro** [hrue/r-inla#30]
On a Windows PC with R 4.1.0 this gives me the error message "Error in qr.coef(a, b) : 'qr' and 'y' must have the same number of rows':

```
Mesh3 <- inla.mesh.2d(grid2, boundary = boundary, max.edge = c(25000, 250000), cutoff = 3000)
matern <- inla.spde2.pcmatern(Mesh3,
                                prior.sigma = c(3000, 0.01),
                                prior.range = c(10, 0.01))
```
(shapefiles here)
[shapefiles.zip](https://github.com/hrue/r-inla/files/6514288/shapefiles.zip)

....I got the same error when running in R4.0.3. So I tried the same code with the same shapefiles on my laptop running R4.0.3 with no error message...and upon updating the laptop to R 4.1.0, it still worked(!).

I'm on the same rgdal version (1.2-23) on both computers, spdep is 1.1-7 and rgeos 0.5-5. I get the same error on the desktop using both RStudio 1.3.1093 and 1.4.1106.

Best wiehs,
Michael

**Comment by finnlindgren** [comment]
Please post the actual code needed to load the data as well, i.e. complete working code.

**Comment by finnlindgren** [comment]
Also, please clarify (e.g. by supplying the sessionInfo() outputs for each case) where it works and where it doesn't. There are laptops with Windows, Linux, and Mac...  But it looks like you're saying that it fails on Windows only, and regardless of the R version?

**Comment by gaiteiro** [comment]
The code is:

```
setwd("C:/Users/MSmith/Documents/R/data/INLA")
grid2 <- readOGR(paste(getwd(), "/PNG_grid_excl_nullrows_singlepart.shp", sep = ""))
boundary <- readOGR(paste(getwd(), "/PNG_boundary_10km_buffer_UTM_ex_BGV.shp", sep = ""))
Mesh3 <- inla.mesh.2d(grid2, boundary = boundary, max.edge = c(25000, 250000), cutoff = 3000)
matern <- inla.spde2.pcmatern(Mesh3,
                                prior.sigma = c(3000, 0.01),
                                prior.range = c(10, 0.01))
```

Both PCs are Windows so I'm not sure about other platforms. Session info for where it works:
R version 4.1.0 (2021-05-18)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19042)

Matrix products: default

locale:
[1] LC_COLLATE=English_United Kingdom.1252  LC_CTYPE=English_United Kingdom.1252    LC_MONETARY=English_United Kingdom.1252
[4] LC_NUMERIC=C                            LC_TIME=English_United Kingdom.1252    

attached base packages:
[1] parallel  stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] rgeos_0.5-5   rgdal_1.5-23  INLA_21.02.23 foreach_1.5.1 Matrix_1.3-3  RANN_2.6.1    spdep_1.1-7   sf_0.9-8      spData_0.3.8 
[10] sp_1.4-5     

loaded via a namespace (and not attached):
 [1] tinytex_0.31       gtools_3.8.2       tidyselect_1.1.1   xfun_0.23          purrr_0.3.4        splines_4.1.0      lattice_0.20-44   
 [8] colorspace_2.0-1   vctrs_0.3.8        generics_0.1.0     expm_0.999-6       utf8_1.2.1         rlang_0.4.11       e1071_1.7-6       
[15] pillar_1.6.1       glue_1.4.2         DBI_1.1.1          lifecycle_1.0.0    inlabru_2.3.1      munsell_0.5.0      gtable_0.3.0      
[22] raster_3.4-10      codetools_0.2-18   coda_0.19-4        knitr_1.33         class_7.3-19       fansi_0.4.2        Rcpp_1.0.6        
[29] KernSmooth_2.23-20 scales_1.1.1       classInt_0.4-3     gdata_2.18.0       deldir_0.2-10      ggplot2_3.3.3      gmodels_2.18.1    
[36] dplyr_1.0.6        grid_4.1.0         cli_2.5.0          tools_4.1.0        LearnBayes_2.15.1  magrittr_2.0.1     proxy_0.4-25      
[43] tibble_3.1.2       crayon_1.4.1       pkgconfig_2.0.3    MASS_7.3-54        ellipsis_0.3.2     iterators_1.0.13   R6_2.5.0          
[50] boot_1.3-28        units_0.7-1        nlme_3.1-152       compiler_4.1.0   

And session info for where it doesn't work:

R version 4.1.0 (2021-05-18)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19042)

Matrix products: default

locale:
[1] LC_COLLATE=English_United Kingdom.1252  LC_CTYPE=English_United Kingdom.1252   
[3] LC_MONETARY=English_United Kingdom.1252 LC_NUMERIC=C                           
[5] LC_TIME=English_United Kingdom.1252    

attached base packages:
[1] parallel  stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] inlabru_2.3.1 ggplot2_3.3.3 rgeos_0.5-5   rgdal_1.5-23  INLA_21.05.02 foreach_1.5.1 Matrix_1.3-3  RANN_2.6.1   
 [9] spdep_1.1-7   sf_0.9-8      spData_0.3.8  sp_1.4-5     

loaded via a namespace (and not attached):
 [1] tinytex_0.31       gtools_3.8.2       tidyselect_1.1.1   xfun_0.23          purrr_0.3.4        splines_4.1.0     
 [7] lattice_0.20-44    colorspace_2.0-1   vctrs_0.3.8        generics_0.1.0     expm_0.999-6       utf8_1.2.1        
[13] rlang_0.4.11       e1071_1.7-6        pillar_1.6.1       withr_2.4.2        glue_1.4.2         DBI_1.1.1         
[19] lifecycle_1.0.0    MatrixModels_0.5-0 munsell_0.5.0      gtable_0.3.0       raster_3.4-10      codetools_0.2-18  
[25] coda_0.19-4        class_7.3-19       fansi_0.4.2        Rcpp_1.0.6         KernSmooth_2.23-20 scales_1.1.1      
[31] classInt_0.4-3     gdata_2.18.0       deldir_0.2-10      gmodels_2.18.1     dplyr_1.0.6        grid_4.1.0        
[37] tools_4.1.0        LearnBayes_2.15.1  magrittr_2.0.1     proxy_0.4-25       tibble_3.1.2       crayon_1.4.1      
[43] pkgconfig_2.0.3    ellipsis_0.3.2     MASS_7.3-54        iterators_1.0.13   assertthat_0.2.1   R6_2.5.0          
[49] boot_1.3-28        units_0.7-1        nlme_3.1-152       compiler_4.1.0    

...hope that helps.

**Comment by finnlindgren** [comment]
So it fails on your Windows "PC", but not on your windows "laptop"?

**Comment by finnlindgren** [comment]
Full code:
```
library(INLA)
library(rgdal)
grid2 <- readOGR(here::here("PNG_grid_excl_nullrows_singlepart.shp"))
boundary <- readOGR(here::here("PNG_boundary_10km_buffer_UTM_ex_BGV.shp"))
Mesh3 <- inla.mesh.2d(grid2, boundary = boundary, max.edge = c(25000, 250000), cutoff = 3000)
matern <- inla.spde2.pcmatern(Mesh3,
                              prior.sigma = c(3000, 0.01),
                              prior.range = c(10, 0.01))
```
runs fine on my Linux system. Since it runs on one of your Windows systems but not the other, my initial guess seems the most likely: that it's a numerical issue and that the qr error is some kind of symptom of that, since the mesh looks likely to have near-singularities. Is there a strong reason for needing to use `grid2` as input points, or would it be enough to specify the boundary? Without `grid2` as input, the mesh seems less problematic:
```
Mesh4 <- inla.mesh.2d(boundary = boundary, max.edge = c(25000, 250000), cutoff = 3000)
```
Another issue may also be the units; converting from metres to kilometres should improve the numerics as well.

**Comment by finnlindgren** [comment]
Also, range 10 is several orders of magnitude smaller than the inner `max.edge` setting, 25000 (as well as the cutoff value 3000). This may also be a cause of the numerical problem.

**Comment by gaiteiro** [comment]
Thanks for your advice....I'm quite new to INLA and was going to try and use inlabru lgcp to model environmental variables at the points of the grid, which are species record locations. Or at least to get an idea of the spatial effect with something like:
```
comp <- coordinates ~ mySmooth(coordinates, model = matern) +  Intercept(1)
fit <- lgcp(comp, grid2, samplers = boundary, domain = list(coordinates = Mesh3))
```
I will try changing the units and range. I  I suppose this is more of a discussion for the INLA google group so I should probably close the issue and follow up there.

**Comment by finnlindgren** [comment]
For point process models, it can be _worse_ to use the observed points as mesh input, since that biases the likelihood approximation. So if you remove that, you might solve both problems.
For georeferenced modelling, i.e. modelling a value at each point, as opposed to modelling the point locations themselves (which is what the lgcp does), then there is sometimes value in useing the points as mesh input, but it does sometimes lead to numerical issues.
I would recommend starting by converting to km. Easiest to convert the geometries before starting to build the mesh etc.
inlabru has helper functions for this; see `?fm_crs_set_lengthunit` for how to convert the crs, and then call `fm_spTransform()` to do the conversion.
The `fm_sp_get_crs()` function can be used to extract the CRS information.

**Comment by finnlindgren** [comment]
Secondly, since the mySmooth model operates on the log-scale, the value 3000 as a "unlikely large sd" is also much larger than reasonable. Even a value as small as 3 might be too large, so perhaps `prior.sigma=c(1,0.01)` may be more reasonable.

**Comment by gaiteiro** [comment]
This is really useful advice, thank you - I'm afraid I fall over a bit between online tutorials (with not very much background) and the documentation pages (for those with more advanced coding skills than mine). 

I'll carry on experimenting, so far tried:
```
mycrs <- fm_sp_get_crs(grid2)
mycrs2 <- fm_crs_set_lengthunit(mycrs, "km")
grid3 <- inla.spTransform(grid2, crs0 = mycrs, crs2 = mycrs2)
```
...but get an error "Invalid target CRS for SpatialPointsDataFrame". Looking at mycrs2 it's clearly incorrect. I thought this might work:

`grid3 <- inla.spTransform(grid2, crs0 = "+proj=utm +zone=54 +south +ellps=WGS72 +units=m +no_defs", crs1 = "+proj=utm +zone=54 +south +ellps=WGS72 +units=km +no_defs")`

...but that returned the same "Invalid target CRS for SpatialPointsDataFrame". Maybe I should try and convert it to WGS84 and see if I can work in degrees...

**Comment by finnlindgren** [comment]
the inlabru fm_* functions may have bug fixes that aren't in the inla-versions, so fm_spTransform may work better.
And "clearly incorrect" is likely not true; proj4 strings can't always be generated for the new proj6 wkt style specifications; you can go from proj4 to proj6, but not always back, and setting the units is one of those cases where rgdal has trouble figuring out what the proj4 would be. as long as the wkt part is ok, it should work. So the problem is likely with inla.spTransform. Let me know if fm_spTransform() works better.

**Comment by finnlindgren** [comment]
Yeah, I've been planning on writing a tutorial for how to properly handle the new PROJ6 stystem when using inla, but haven't had time yet...

**Comment by gaiteiro** [comment]
Oh that's odd, I get 'could not find function "fm_spTransform" and if I try inlabru::fm_spTransform I'm told it's 'not an exported object form namespace:inlabru'.

I don't envy you the task of writing about projections! It's an admirable idea though. For jobbing ecologists PROJ4 strings are the stuff of nightmares.

**Comment by finnlindgren** [comment]
That's odd! It should be in your inlabru version. Perhaps reinstall inlabru?

**Comment by gaiteiro** [comment]
No luck I'm afraid...looking at https://sites.google.com/inlabru.org/inlabru, I made sure to update INLA to the testing version but that didn't help. I also tried the latest development version but failed with the installation (although I generally have no trouble with 'remotes').  I had a look on RStudio my old laptop, it too can't find fm_spTransform. On the whole, inlabru seems to be functioning - e.g I can run the 02_random_fields.R tutorial code.

**Comment by finnlindgren** [comment]
Please confirm that you have at least the latest cran version of inlabru, or the latest github dev version? fm_spTransform is an S3 method that isn't exported separately, which may explain the lack of direct inlabru::fm_spTransform access to the function itself, but the documentation should be available, and the function should be useable, `?inlabru::fm_spTransform`. The function itself should exist as `inlabru:::fm_spTransform`. Make sure you've done `library(sp)`, in case that affects the S3 method dispatch. Many other S3methods are also explicitly exported; not sure why this one isn't.

**Comment by finnlindgren** [comment]
Try the latest inlabru devel version from github; I just added an explicit `@export` for `fm_spTransform`, which should solve the problem.

**Comment by gaiteiro** [comment]
Thanks! That gets fm_spTransform recognized, although I haven't quite got things going (see below for outputs).

I thought I'd try fitting a model on my old laptop (still in metres admittedly) as follows, with the boundary only as you suggested:
```
Mesh4 <-inla.mesh.2d(boundary = boundary, max.edge=c(25000, 250000), cutoff = 3000)
matern <- inla.spde2.pcmatern(Mesh4,
                              prior.sigma = c(1, 0.01),
                              prior.range = c(3000, 0.01))
comp <- coordinates ~ mySmooth(coordinates, model = matern) +  Intercept(1)
fit <- lgcp(comp, data = boundary, domain = list(coordinates = Mesh4))
```
...but I suppose that's the wrong syntax for the lgcp? At that point it returns "Error in dimnames(x@bbox) <- '*vtmp*' : length of 'dimnames' [1] not equal to array extent" - sorry probably some beginner's error there...

Anyway: for fm_spTransform:

> mycrs
CRS arguments:
 +proj=utm +zone=54 +south +ellps=WGS72 +units=m +no_defs 

> mycrs2
CRS arguments:
 +proj=tmerc +lat_0=0 +lon_0=141 +k=0.9996 +x_0=500000 +y_0=10000000 +ellps=WGS72 +units=km +no_defs 

> class(grid2)
[1] "SpatialPointsDataFrame"
attr(,"package")
[1] "sp"

> grid3 <- fm_spTransform(grid2, crs0 = mycrs, crs1 = mycrs2)
Error in fm_spTransform.SpatialPointsDataFrame(grid2, crs0 = mycrs, crs1 = mycrs2) : 
  Invalid target CRS for sp::SpatialPointsDataFrame

> class(boundary)
[1] "SpatialPolygonsDataFrame"
attr(,"package")
[1] "sp"

> boundary2 <- fm_spTransform(boundary, crs0 = mycrs, crs1 = mycrs2)
Error in fm_crs_bounds_check(x, bounds) : 
  inherits(x, "matrix") is not TRUE
In addition: Warning messages:
1: In showSRID(SRS_string, format = "PROJ", multiline = "NO", prefer_proj = prefer_proj) :
  Discarded ellps unknown in Proj4 definition: +proj=longlat +R=1 +no_defs +type=crs
2: In showSRID(SRS_string, format = "PROJ", multiline = "NO", prefer_proj = prefer_proj) :
  Discarded datum unknown in Proj4 definition
3: In showSRID(SRS_string, format = "PROJ", multiline = "NO", prefer_proj = prefer_proj) :
  Discarded ellps unknown in Proj4 definition: +proj=longlat +R=6378135 +no_defs +type=crs
4: In showSRID(SRS_string, format = "PROJ", multiline = "NO", prefer_proj = prefer_proj) :
  Discarded datum unknown in Proj4 definition
5: In showSRID(SRS_string, format = "PROJ", multiline = "NO", prefer_proj = prefer_proj) :
  Discarded ellps unknown in Proj4 definition: +proj=longlat +R=6378135 +no_defs +type=crs
6: In showSRID(SRS_string, format = "PROJ", multiline = "NO", prefer_proj = prefer_proj) :
  Discarded datum unknown in Proj4 definition
7: In showSRID(SRS_string, format = "PROJ", multiline = "NO", prefer_proj = prefer_proj) :
  Discarded ellps unknown in Proj4 definition: +proj=geocent +R=1 +units=m +no_defs +type=crs
8: In showSRID(SRS_string, format = "PROJ", multiline = "NO", prefer_proj = prefer_proj) :
  Discarded datum unknown in Proj4 definition

**Comment by finnlindgren** [comment]
The `lgcp` syntax looks fine; it seems to be confused about names of the coordinate variables. Make sure they have names (via `coordnames()` I think).

For the rest, I to run it here, but can't do that today. Hopefully tomorrow.

**Comment by gaiteiro** [comment]
Thank you and no hurry of course - on the contrary, I hugely appreciate your patience so far.

**Comment by gaiteiro** [comment]
Perhaps I didn't need to have recourse to fm_spTransform as this worked:

`boundary= spTransform(boundary, "+proj=utm +zone=54 +south +ellps=WGS72 +units=km +no_defs")`

Looking at my mesh, I saw that CRS was NULL - it hadn't acquired the CRS of the SpatialPolygonsDataFrame boundary.

....so having transformed the dataframes to km units, the fit seems to work after I add the following argument to inla.mesh.2d: `crs="+proj=utm +zone=54 +south +ellps=WGS72 +units=km +no_defs"`
